File: tikzit-1/src/common/Node.m
//
//  Node.m
//  TikZiT
//  
//  Copyright 2010 Aleks Kissinger. All rights reserved.
//  
//  
//  This file is part of TikZiT.
//  
//  TikZiT is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//  
//  TikZiT is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//  
//  You should have received a copy of the GNU General Public License
//  along with TikZiT.  If not, see <http://www.gnu.org/licenses/>.
//  

#import "Node.h"

#import "Shape.h"


@implementation Node

- (id)initWithPoint:(NSPoint)p {
	self = [super init];
	if (self) {
		data = [[GraphElementData alloc] init];
		style = nil;
		label = @"";
		point = p;
	}
	return self;
}

- (id)init {
	return [self initWithPoint:NSMakePoint(0.0f, 0.0f)];
}

- (id)copyWithZone:(NSZone*)z {
	Node *cp = [[Node allocWithZone:z] init];
	[cp setPropertiesFromNode:self];
	return cp;
}

- (void)dealloc {
#if ! __has_feature(objc_arc)
	[name release];
	[style release];
	[data release];
	[label release];
	[super dealloc];
#endif
}

- (Shape*) shape {
    if (style) {
        return [Shape shapeForName:[style shapeName]];
    } else {
        return nil;
    }
}

- (Transformer*) shapeTransformerFromTransformer:(Transformer*)t {
	// we take a copy to keep the reflection attributes
#if ! __has_feature(objc_arc)
    Transformer *transformer = [[t copy] autorelease];
#else
    Transformer *transformer = [t copy];
#endif
    NSPoint screenPos = [t toScreen:point];
    [transformer setOrigin:screenPos];
    float scale = [t scale];
    if (style) {
        scale *= [style scale];
    }
    [transformer setScale:scale];
    return transformer;
}

- (Transformer*) shapeTransformer {
    float scale = 1.0f;
    if (style) {
        scale = [style scale];
	}
    return [Transformer transformerWithOrigin:point andScale:scale];
}

- (NSRect) boundsUsingShapeTransform:(Transformer*)shapeTrans {
	//if (style) {
		return [shapeTrans rectToScreen:[[self shape] boundingRect]];
	/*} else {
		NSRect r = NSZeroRect;
		r.origin = [shapeTrans toScreen:[self point]];
		return r;
	}*/
}

- (NSRect) boundingRect {
	return [self boundsUsingShapeTransform:[self shapeTransformer]];
}

- (BOOL)attachStyleFromTable:(NSArray*)styles {
#if __has_feature(objc_arc)
    NSString *style_name = [data propertyForKey:@"style"];
#else
    NSString *style_name = [[[data propertyForKey:@"style"] retain] autorelease];
#endif
	
	[self setStyle:nil];
	
	// 'none' is a reserved style
	if (style_name == nil || [style_name isEqualToString:@"none"]) return YES;
	
	for (NodeStyle *s in styles) {
		if ([[s name] compare:style_name]==NSOrderedSame) {
			[self setStyle:s];
			return YES;
		}
	}
	
	// if we didn't find a style, fill in a default one
	[self setStyle:[NodeStyle defaultNodeStyleWithName:style_name]];
	return NO;
}

- (void)updateData {
	if (style == nil) {
		[data setProperty:@"none" forKey:@"style"];
	} else {
		[data setProperty:[style name] forKey:@"style"];
	}
}

- (void)setPropertiesFromNode:(Node*)nd {
	[self setPoint:[nd point]];
	[self setStyle:[nd style]];
	[self setName:[nd name]];
	[self setData:[nd data]];
	[self setLabel:[nd label]];
}

+ (Node*)nodeWithPoint:(NSPoint)p {
#if __has_feature(objc_arc)
    return [[Node alloc] initWithPoint:p];
#else
    return [[[Node alloc] initWithPoint:p] autorelease];
#endif
}

+ (Node*)node {
#if __has_feature(objc_arc)
    return [[Node alloc] init];
#else
    return [[[Node alloc] init] autorelease];
#endif
}


// perform a lexicographic ordering (-y, x) on coordinates.
- (NSComparisonResult)compareTo:(id)nd {
	Node *node = (Node*)nd;
	if (point.y > [node point].y) return NSOrderedAscending;
	else if (point.y < [node point].y) return NSOrderedDescending;
	else {
		if (point.x < [node point].x) return NSOrderedAscending;
		else if (point.x > [node point].x) return NSOrderedDescending;
		else return NSOrderedSame;
	}
}

@synthesize name;
@synthesize label;
@synthesize point;

@synthesize data;
- (void) insertObject:(GraphElementProperty*)gep
		inDataAtIndex:(NSUInteger)index {
	[data insertObject:gep atIndex:index];
}
- (void) removeObjectFromDataAtIndex:(NSUInteger)index {
	[data removeObjectAtIndex:index];
}
- (void) replaceObjectInDataAtIndex:(NSUInteger)index
						 withObject:(GraphElementProperty*)gep {
	[data replaceObjectAtIndex:index withObject:gep];
}

- (NodeStyle*)style {
	return style;
}

- (void)setStyle:(NodeStyle *)st {
	if (style != st) {
#if __has_feature(objc_arc)
        style = st;
#else
        NodeStyle *oldStyle = style;
        style = [st retain];
        [oldStyle release];
#endif
	}
	[self updateData];
}

@end

// vi:ft=objc:ts=4:noet:sts=4:sw=4

File: src/data/node.cpp
/*
    TikZiT - a GUI diagram editor for TikZ
    Copyright (C) 2018 Aleks Kissinger

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include "node.h"
#include "tikzit.h"

#include <QDebug>

Node::Node(QObject *parent) : QObject(parent), _tikzLine(-1)
{
    _data = new GraphElementData();
    _style = noneStyle;
    _data->setProperty("style", "none");
}

Node::~Node()
{
    delete _data;
}

Node *Node::copy() {
    Node *n1 = new Node();
    n1->setName(name());
    n1->setData(data()->copy());
    n1->setPoint(point());
    n1->setLabel(label());
    n1->attachStyle();
    n1->setTikzLine(tikzLine());
    return n1;
}

QPointF Node::point() const
{
    return _point;
}

void Node::setPoint(const QPointF &point)
{
    _point = point;
}

QString Node::name() const
{
    return _name;
}

void Node::setName(const QString &name)
{
    _name = name;
}

QString Node::label() const
{
    return _label;
}

void Node::setLabel(const QString &label)
{
    _label = label;
}

GraphElementData *Node::data() const
{
    return _data;
}

void Node::setData(GraphElementData *data)
{
    delete _data;
    _data = data;
}

QString Node::styleName() const
{
    return _data->property("style");
}

void Node::setStyleName(const QString &styleName)
{
    _data->setProperty("style", styleName);
}

void Node::attachStyle()
{
    QString nm = styleName();
    if (nm == "none") _style = noneStyle;
    else _style = tikzit->styles()->nodeStyle(nm);
}

NodeStyle *Node::style() const
{
    return _style;
}

bool Node::isBlankNode()
{
    return styleName() == "none";
}

int Node::tikzLine() const
{
    return _tikzLine;
}

void Node::setTikzLine(int tikzLine)
{
    _tikzLine = tikzLine;
}
